src/box/lua/config/utils/expression_lexer.c

```diff
@@ -0,0 +1,280 @@
+		 *  - 'continue' rechecks the state without
+		 *    skipping the current character.
+		 */
+		switch (state) {
+		case START:
+			/*
+			 * Stop parsing at end of the input
+			 * string.
+			 */
+			if (*s == '\0')
+				break;
+			/* Skip spaces. */
+			if (isspace(*s))
+				break;
+			/* Variable. */
+			if (isalpha(*s) || *s == '_') {
+				state = VARIABLE;
+				token_start = s;
+				break;
+			}
+			/* Version literal. */
+			if (isdigit(*s)) {
+				state = VERSION_LITERAL_DIGIT;
+				token_start = s;
+				literal_components = 1;
+				break;
+			}
+			/* Single character operator. */
+			if (*s == '(' || *s == ')') {
+				PUSH_TOKEN("grouping", s, s);
+				break;
+			}
+			/* Single or two character operator. */
+			if (*s == '>' || *s == '<') {
+				state = COMPARE;
+				break;
+			}
+			/* Two character operators. */
+			if (*s == '&' || *s == '|') {
+				state = LOGICAL;
+				break;
+			}
+			if (*s == '!' || *s == '=') {
+				state = EQUALITY;
+				break;
+			}
+			return ERROR("invalid token");
+		case VARIABLE:
+			/*
+			 * Consume a-z, A-Z, 0-9 and underscore.
+			 */
+			if (isalnum(*s) || *s == '_')
+				break;
+			/*
+			 * The series of the variable characters
+			 * has been ended.
+			 *
+			 * Push the token and verify that the next
+			 * character is a separator.
+			 */
+			PUSH_TOKEN("variable", token_start, s - 1);
+			token_start = NULL;
+			state = NEEDS_SEPARATOR;
+			continue;
+		case VERSION_LITERAL_DIGIT:
+			/* Consume 0-9. */
+			if (isdigit(*s))
+				break;
+			/* Consume full stop. */
+			if (*s == '.') {
+				++literal_components;
+				state = VERSION_LITERAL_FULL_STOP;
+				break;
+			}
+			/*
+			 * The series of digits and full stops has
+			 * been ended.
+			 *
+			 * Verify amount of components in the version
+			 * literal.
+			 */
+			if (literal_components != 3)
+				return ERROR("invalid version literal: "
+					     "expected 3 components, got %d",
+					     literal_components);
+			/*
+			 * Push the token and verify that the next
+			 * character is a separator.
+			 */
+			PUSH_TOKEN("version_literal", token_start, s - 1);
+			token_start = NULL;
+			literal_components = 0;
+			state = NEEDS_SEPARATOR;
+			continue;
+		case VERSION_LITERAL_FULL_STOP:
+			/* Consume 0-9. */
+			if (isdigit(*s)) {
+				state = VERSION_LITERAL_DIGIT;
+				break;
+			}
+			/*
+			 * Forbid a second full stop in a row.
+			 *
+			 * Forbid a version literal ending with a
+			 * full stop.
+			 */
+			return ERROR("invalid token");
+		case NEEDS_SEPARATOR:
+			/*
+			 * The end of input, a space, an operator
+			 * symbol are the separators.
+			 */
+			if (*s == '\0' || isspace(*s) ||
+			    *s == '(' || *s == ')' ||
+			    *s == '>' || *s == '<' ||
+			    *s == '&' || *s == '|' ||
+			    *s == '!' || *s == '=') {
+				state = START;
+				continue;
+			}
+			return ERROR("invalid token");
+		case COMPARE:
+			/* Push >= or <=. */
+			if (*s == '=') {
+				PUSH_TOKEN("operation", s - 1, s);
```

Comment:
It fails on gcc 4.8.5. I'm investigating it.

```c
/build/usr/src/debug/tarantool-3.0.0~beta1.78.dev/src/box/lua/config/utils/expression_lexer.c: In function ‘luaT_expression_lexer_split’:
/build/usr/src/debug/tarantool-3.0.0~beta1.78.dev/src/box/lua/config/utils/expression_lexer.c:79:48: error: integer overflow in expression [-Werror=overflow]
  lua_pushlstring(L, (token_start), (token_end) - (token_start) + 1); \
                                                ^
/build/usr/src/debug/tarantool-3.0.0~beta1.78.dev/src/box/lua/config/utils/expression_lexer.c:212:5: note: in expansion of macro ‘PUSH_TOKEN’
     PUSH_TOKEN("operation", s - 1, s);
     ^
```

Reply:
It seems it is due to https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61240.

Reply:
The following workaround works for me.

```diff
diff --git a/src/box/lua/config/utils/expression_lexer.c b/src/box/lua/config/utils/expression_lexer.c
index 5e61d68..faa7880 100644
--- a/src/box/lua/config/utils/expression_lexer.c
+++ b/src/box/lua/config/utils/expression_lexer.c
@@ -73,10 +73,17 @@
 	luaL_error(L, "Expression parsing error at line %d, column %d: " fmt,	\
 		   line, column, ##__VA_ARGS__)
 #define PUSH_TOKEN(token_type, token_start, token_end) do {			\
+	/*									\
+	 * Workaround false-positive integer overflow warning on		\
+	 * GCC 4.8.5.								\
+	 *									\
+	 * https://gcc.gnu.org/bugzilla/show_bug.cgi?id=61240			\
+	 */									\
+	const char *token_start_var = (token_start);				\
 	lua_createtable(L, 0, 2);						\
 	lua_pushliteral(L, token_type);						\
 	lua_setfield(L, -2, "type");						\
-	lua_pushlstring(L, (token_start), (token_end) - (token_start) + 1);	\
+	lua_pushlstring(L, (token_start), (token_end) - token_start_var + 1);	\
 	lua_setfield(L, -2, "value");						\
 	lua_rawseti(L, -2, lua_objlen(L, -2) + 1);				\
 } while (0)
```

I'm going to apply it and look again on CI.

Reply:
Added the workaround.